ompi/mca/osc/rdma/osc_rdma_passive_target.c

```diff
@@ -309,6 +442,53 @@ int ompi_osc_rdma_unlock_atomic (int target, ompi_win_t *win)
     return ret;
 }
 
+int ompi_osc_rdma_iunlock_atomic (int target, ompi_win_t *win)
+{
+    ompi_osc_rdma_module_t *module = GET_MODULE(win);
+    ompi_osc_rdma_peer_t *peer;
+    ompi_osc_rdma_sync_t *lock;
+    int ret = OMPI_SUCCESS;
+
+    OPAL_THREAD_LOCK(&module->lock);
+
+    OSC_RDMA_VERBOSE(MCA_BASE_VERBOSE_TRACE, "iunlock: %d, %s", target, win->w_name);
+
+    lock = ompi_osc_rdma_module_lock_find (module, target, &peer);
+    if (OPAL_UNLIKELY(NULL == lock)) {
+        OSC_RDMA_VERBOSE(MCA_BASE_VERBOSE_INFO, "target %d is not locked in window %s",
+                         target, win->w_name);
+        OPAL_THREAD_UNLOCK(&module->lock);
+        return OMPI_ERR_RMA_SYNC;
+    }
+
+    ompi_osc_rdma_module_lock_remove (module, lock);
+
+    /* finish all outstanding fragments */
+    if(ompi_osc_rdma_sync_rdma_icomplete (lock) == false){
+        return OMPI_ERR_RMA_SYNC;
+    }
+
+    if (!(lock->sync.lock.assert & MPI_MODE_NOCHECK)) {
+        ret = ompi_osc_rdma_iunlock_atomic_internal (module, peer, lock);
```

Comment:
What happens when `ompi_osc_rdma_iunlock_atomic_internal` returns `false` (or our new `OMPI_ERR_RMA_NB_PENDING` error code), i.e. the lock was not actually released?

This situation needs additional code to return `OMPI_ERR_RMA_NB_PENDING` from this function *before* destroying the `lock` object.

Reply:
Do you mean ```ompi_osc_rdma_iunlock_atomic``` ? ```*_internal``` always returns ```OMPI_SUCCESS```

Reply:
Changes to that `_internal` function will make it return `OMPI_ERR_RMA_NB_PENDING` when it tries the action but does not succeed. By "that `_internal` function, I mean: this code will call `_tryunlock_` rather than `_iunlock_` as in the current version (or `_unlock_` as in the original blocking version). The `_tryunlock_` function can return `OMPI_ERR_RMA_NB_PENDING` if the try didn't work.